[tap-frontapp] - fix libraries and dependabot issues

romilqlik opened the pull request:
# Description of change
Updated FrontApp tap with Schema fixes and unit tests.
Fixed and streamlined 'init.py by ensuring all streams and their fields have proper metadata with 'selected: true' at both stream and field levels (breadcrumb[]).
Added .circleci/config.yml to trigger CircleCI pipeline


# Manual QA steps
 Ran Discovery
Ran full sync
Validated against FrontApp API docs.
Unit test executed
 
# Risks
 None
 
# Rollback steps
 - revert this branch

#### AI generated code
https://internal.qlik.dev/general/ways-of-working/code-reviews/#guidelines-for-ai-generated-code
- [ ] this PR has been written with the help of GitHub Copilot or another generative AI tool

gl-romil committed:
[SAC-27536] Frontapp App libary changes

RushiT0122 commented on tests/unittests/test_client.py:
```diff
@@ -0,0 +1,81 @@
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Add new line at end of the file.

RushiT0122 commented on setup.py:
```diff
@@ -10,11 +10,11 @@
-        "backoff==1.3.2",
-        "requests==2.31.0",
+        "singer-python==6.1.1",
+        "pendulum==3.1.0",
+        "ratelimit==2.2.1",
```
@rdeshmukh15 need to get these dependancies whitelisted.

romilqlik replied:
Can you provide:
Which older versions of pendulum and rate limit are currently whitelisted?
Should we revert to those older approved versions for now or wait for the current versions to be approved.

RushiT0122 commented on tap_frontapp/__init__.py:
```diff
@@ -23,40 +23,117 @@
+        # Stream-level metadata select the stream
+        metadata.append({
+            "metadata": {
+                "selected": True  # Make sure to select every stream
```
@gl-romil `selected` is a non-discoverable metadata which can't written or read by the tap. Please refer [getting-started documentation](https://github.com/singer-io/getting-started/blob/1b94abc1f5aecbeef0322cae1a4a8140ad2de8b6/docs/DISCOVERY_MODE.md#metadata). Also I find some required discoverable metadata is missing from the catalog, refer docs for more details.

romilqlik replied:
Replaced 'selected' with discoverable metadata fields.
Added missing 'inclusion', inclusion-reason' , and 'selected-by-default'.

RushiT0122 commented on tap_frontapp/__init__.py:
```diff
@@ -23,40 +23,117 @@
+    # Creating a dict to change before converting
+    catalog_dict = catalog.to_dict()
+    
+    required_streams = [
```
You can use `streams.METRIC_API_DESCRIPTION_KEY` instead of introducing new variable which already defines these stream names to required.

romilqlik replied:
Replaced static stream list with streams.METRIC_API_DESCRIPTION_KEY.keys()

RushiT0122 reviewed (changes requested):
Requested changes in `discover()` which may need changes in the other tap implementation as well as in the tests. Please take of those as well.

romilqlik reviewed (commented):
All review comments are addressed.

RushiT0122 commented on setup.py:
```diff
@@ -25,4 +25,4 @@
     },
     include_package_data=True
-)
+)
```
I see new line is still missing.

romilqlik reviewed (commented):
Kindly review

romilqlik committed:
[SAC-27535] [tap-frontapp] - fix extraction failures (#28)

* [SAC-27536] Refactored discovery and sync- logic

* [SAC-27535] Final clean up: discovery fixes,init and sync

* Update discover.py

Removed type annotations from the file as per review

* Update schemas.py

Removed type annotations from file as per review

* Update sync.py

Removed type annotations as per review

* Update sync.py

* Update __init__.py

Removed type annotations as per comment

* Update __init__.py

* Update __init__.py

Removed unnecessary header as per review

* Update __init__.py

* Update __init__.py

Removed header comment

* Update sync.py

Removed file level docstring

* Update test_client.py

Add EOF

---------

Co-authored-by: gl-romil <129256325+gl-romil@users.noreply.github.com>

RushiT0122 commented on tap_frontapp/discover.py:
```diff
@@ -0,0 +1,38 @@
+def discover():
+    """Run the discovery mode, prepare the catalog file and return the catalog."""
+    schemas, field_metadata = get_schemas()
+    print("Schemas loaded:", schemas.keys())
```
Please replace `print()` but `LOGGER.info()`

RushiT0122 replied:
Also we need to do credential validation as we discussed for `tap-braintree`.

romilqlik replied:
This is Done, Kindly check

romilqlik committed:
Update discover.py

Added LOGGER.info

romilqlik committed:
Update __init__.py

Credentials Validation

romilqlik reviewed (commented):
Created validate_credential function in init.py file
Replaces print with LOGGER

romilqlik committed:
Update setup.py

romilqlik committed:
Update __init__.py

romilqlik committed:
Update discover.py

Moved validate_credentials

romilqlik committed:
Update __init__.py

romilqlik committed:
Update __init__.py

romilqlik committed:
Update discover.py

romilqlik reviewed (commented):
Kindly review

romilqlik commented on tap_frontapp/__init__.py:
```diff
@@ -6,88 +6,61 @@
+            sys.exit(1)
+    except requests.exceptions.RequestException as err:
+        LOGGER.critical("Credential validation failed: %s", str(err))
+        sys.exit(1)
```
This is Done, Kindly review

RushiT0122 commented on tests/unittests/test_client.py:
```diff
@@ -0,0 +1,81 @@
+        with self.assertRaises(ConnectionError):
+            client.get_report_metrics("https://api2.frontapp.com/analytics/reports/xyz")
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Please remove this but not holding PR for this.

romilqlik replied:
I have removed it.

romilqlik committed:
Update test_client.py

romilqlik reviewed (commented):
Removed
if __name__ == "__main__":
+    unittest.main()

romilqlik committed:
Qa bugfixes from sac 27536 (#30)

* Remove non-discoverable inclusion-reason metadata- as per QA feedback

* Add JSON validator step to Circleci as per QA feedback

* Remove selected by-default metadata as per feedback

---------

Co-authored-by: gl-romil <129256325+gl-romil@users.noreply.github.com>

rdeshmukh15 committed:
requests update to 2.32.4

mittal-tushar committed:
selected field removal from schema files (#33)

mittal-tushar committed:
[SAC-27953] - Rate limit change for `get_report_metrics` API (#34)

* rate limit change

* rate limit change

mittal-tushar committed:
Merge branch 'master' of https://github.com/singer-io/tap-frontapp into feature/SAC-27536-frontapp-libraries

mittal-tushar committed:
[SAC-27953] - Usage of `retry-after` header for 429 errors and throttling for `create_report` API (#35)

* RETRY_RATE_LIMIT change

* period set to 3

rdeshmukh15 committed:
Merge branch 'master' into feature/SAC-27536-frontapp-libraries

rdeshmukh15 committed:
Update config.yml
